File: 04-practice/ex04-bank/solution.py
class BankAccount:
    def __init__(self, owner, opening=0):
        self.owner = owner
        self._balance = opening

    @property
    def balance(self):
        return self._balance

    def deposit(self, amount):
        if amount <= 0:
            raise ValueError("amount must be positive")
        self._balance += amount

    def withdraw(self, amount):
        if amount <= 0:
            raise ValueError("amount must be positive")
        if amount > self._balance:
            raise ValueError("insufficient funds")
        self._balance -= amount


def assert_(cond, label):
    print(f"OK   {label}" if cond else f"FAIL {label}")


a = BankAccount("Ada", 100)
a.deposit(40)
assert_(a.balance == 140, "deposit")
a.withdraw(20)
assert_(a.balance == 120, "withdraw")
threw = False
try:
    a.withdraw(1000)
except Exception:
    threw = True
assert_(threw, "overdraft raises")

File: 04-practice/ex04-bank/solution.js
class BankAccount {
  #balance = 0;

  constructor(owner, opening = 0) {
    this.owner = owner;
    this.#balance = opening;
  }

  deposit(amount) {
    if (amount <= 0) throw new Error("amount must be positive");
    this.#balance += amount;
  }

  withdraw(amount) {
    if (amount <= 0) throw new Error("amount must be positive");
    if (amount > this.#balance) throw new Error("insufficient funds");
    this.#balance -= amount;
  }

  get balance() {
    return this.#balance;
  }
}

function assert(cond, label) {
  console.log(cond ? `OK   ${label}` : `FAIL ${label}`);
}

const a = new BankAccount("Ada", 100);
a.deposit(40);
assert(a.balance === 140, "deposit");
a.withdraw(20);
assert(a.balance === 120, "withdraw");
let threw = false;
try {
  a.withdraw(1000);
} catch {
  threw = true;
}
assert(threw, "overdraft throws");
